gmqtt/aioclient.py

```diff
@@ -0,0 +1,91 @@
+import asyncio
+from .client import Client, Message
+from typing import Optional, Tuple
+
+
+class MqttClientProtocol:
+    """
+    Wraps the normal client in the async/await API
+    """
+
+    def __init__(
+        self, inner_client: Client, loop: Optional[asyncio.AbstractEventLoop] = None
+    ):
+        if loop is None:
+            loop = asyncio.get_event_loop()
+        self.loop = loop
+
+        self.cbclient = inner_client
+        self.message_queue = asyncio.Queue()
```

Comment:
This is dangerous, because if your "reading-code" will be blocked for long time - it will broduce high memory usage and we will get memory overflow. So it will be nice to limit queue by option in __init__ method.

Reply:
Good point. I just looked at the flow control section of the MQTT spec. The broker won't send more messages than the client's send quota, initially set to `receive_maximum`. I imagine this is so the client's buffer doesn't overflow. The quota is incremented by a PUBACK, should we send this only once the message has been delivered to the user? (popped off the queue)

Reply:
There might already be something like this in the codebase. I'll have a look around.

Reply:
Yes, `receive_maximum` may be a good solution, but it's works only for `qos1`, `qos2` not for `qos0`.

>  The quota is incremented by a PUBACK, should we send this only once the message has been delivered to the user?
 
In the current implementation you may choose when send `PUBACK` (by setting `optimistic_acknowledgement` to True or False) - after we received message (but before user process if by `on_message` callback) or after, it will be processed by `on_message` callback and will use result of it for `PUBACK`. 

So in my opinion,  it's good to save this behaviour and allow to user choose when `PUBACK` will be sent.

Reply:
Sweet, I chased down the `optimistic_acknowledgement` setting and found that what you described;  the PUBACK is sent before or after the `on_message` callback has been run. I think I understand the usecase for optimistic acknowledgements now.

Without optimistic acknowledgments, `sub1` will continue to receive messages but will never respond with a PUBACK:
```py
async with gmqtt.connect(...) as client:
    sub1 = client.subscribe(...)
    async for message in client.subscribe(...):
        print(message)
```
I used to think the lack of PUBACK's would prompt a retransmission, but that doesn't seem to be the case. [Section 4.4](http://docs.oasis-open.org/mqtt/mqtt/v5.0/cs02/mqtt-v5.0-cs02.html#_Message_delivery_retry) of the spec says:

> When a Client reconnects with Clean Start set to 0 and a session is present, both the Client and Server MUST resend any unacknowledged PUBLISH packets (where QoS > 0) and PUBREL packets using their original Packet Identifiers. **This is the only circumstance where a Client or Server is REQUIRED to resend messages.** Clients and Servers MUST NOT resend messages at any other time.

I interpret this to mean that the broker won't retransmit packets. PUBACK is a notification of the client "taking ownership" of the message; the point where we are happy to let the broker forget the message. To me, that means once the message is delivered to the user. Optimistic acknowledgements just disable the flow control feature of MQTT (designed to keep our buffer from overflowing), and saves the broker from storing the messages while they sit in our buffer; An extremely light burden for most cases.

Anyway, philosophical rant over. I'm happy to add the option for optimistic acknowledgments, it should just be passing the parameter through to the standard client. I don't think I'd recommend the feature to most people though. Perhaps you can convince me otherwise? :stuck_out_tongue: 


Currently, the `__handle_publish_callback` is handled after `on_message` has returned. https://github.com/wialon/gmqtt/blob/4fb92a9629fc57c885f5e07c46b951c4f45e9572/gmqtt/mqtt/handler.py#L330 

For non-optimistic acknowledgements, the library should send the PUBACK as soon as the message is delivered to the user. 

My wrapper will need a special case where the `__handle_publish_callback` is passed to the `on_message` handler rather than run after it completes.

Subclassing `Client` is one option.